src/main/java/com/redhat/devtools/lsp4ij/operations/codelens/LSPCodeLensProvider.java

```diff
@@ -0,0 +1,181 @@
+            while (!future.isDone()) {
+                ProgressManager.checkCanceled();
+                try {
+                    List<CodeLensData> datas = future.get(25, TimeUnit.MILLISECONDS);
+                    fillCodeLenses(document, datas, codelenses);
+                } catch (TimeoutException | ExecutionException e) {
+                    // Do nothing
+                }
+            }
+        }
+        return codelenses;
+    }
+
+    private static void fillCodeLenses(Document document, List<CodeLensData> datas, List<Pair<Integer, CodeLensData>> codelenses) {
+        for (var codeLensData : datas) {
+            int offset = LSPIJUtils.toOffset(codeLensData.codeLens().getRange().getStart(), document);
+            codelenses.add(Pair.create(offset, codeLensData));
+        }
+    }
+
+    private InlayPresentation toPresentation(
+            @NotNull Editor editor,
+            int offset,
+            @NotNull List<Pair<Integer, CodeLensData>> elements,
+            @NotNull PresentationFactory factory,
+            @NotNull List<CompletableFuture> toResolve,
+            @NotNull CancellationSupport cancellationSupport) {
+        int line = editor.getDocument().getLineNumber(offset);
+        int column = offset - editor.getDocument().getLineStartOffset(line);
+        List<InlayPresentation> presentations = new ArrayList<>();
+        presentations.add(factory.textSpacePlaceholder(column, true));
+        elements.forEach(p -> {
+            cancellationSupport.checkCanceled();
+            CodeLensData codeLensData = p.getSecond();
+            CodeLens codeLens = codeLensData.codeLens();
+            CompletableFuture<CodeLens> resolvedCodeLensFuture = codeLensData.resolvedCodeLensFuture();
+            if (resolvedCodeLensFuture != null) {
+                // Try to get the result of the codelens resolve
+                CodeLens resolved = resolvedCodeLensFuture.getNow(RESOLVE_CODELENS_NOT_READY);
+                if (resolved != null && RESOLVE_CODELENS_NOT_READY != resolved) {
+                    codeLens = resolved;
+                }
+            }
+            if (codeLens.getCommand() != null) {
+                // The codelens is valid, display it
+                InlayPresentation text = factory.smallText(getCodeLensContent(codeLens));
+                if (!hasCommand(codeLens)) {
+                    // No command, create a simple text inlay hint
+                    presentations.add(text);
+                } else {
+                    // Codelens defines a Command, create a clickable inlay hint
+                    final CodeLens resolvedCodeLens = codeLens;
+                    LanguageServer languageServer = p.getSecond().languageServer().getServer();
+                    InlayPresentation clickableText = factory.referenceOnHover(text, (event, translated) ->
+                            executeClientCommand(languageServer, resolvedCodeLens, (Component) event.getSource(), editor.getProject())
```

Comment:
can it be a server command instead?

Reply:
I think you are right, I tried to implement that but it breaks IJ Quarkus command. 

It requires some refactoring. As this PR is already big, please create an issue for that.